### PR TITLE
[Digital Ocean] Code cleanup with no functional modifications

### PR DESCRIPTION
**Pull request by srikiz**
- Using interfaces for better unit testing (haven't added yet, but will create that separately)
- Removing cyclic dependency - moved the code from pkg/resources/digitalocean/cloud.go to upup/pkg/fi/cloudup/do/cloud.go.
- No changes to the functionality, but just moving code so it's easier to maintain going forward.

FYI - @timoreimann

### CODE REVIEW

**Comment by johngmyers** [comment]
Please edit commit history so reviewers don't have to see the back-and-forth. A series of smaller commits is better, making it easier for reviewers to verify that any given refactoring has no functional modification, but the stream should not add an unrelated file and then delete it later.

**Review by timoreimann — APPROVED** [review]
Looks good from a DO perspective. I'll leave the official approval to those that bear the right powers. :)

**Comment by johngmyers** [comment]
@srikiz I'm concerned this might be a bit too big to review in one go. It's easier to review if code is moved in one commit and then cleaned up in a separate commit.

**Comment by srikiz** [comment]
@johngmyers - that sounds like a plan. Once I get the issues resolved, I will close this pull request, and create a new one with 2 commits as you suggested. I'll just move the file in the first commit and keep the rest of the changes in a different commit.

**Comment by johngmyers** [comment]
@srikiz alternatively you could force-push to this PR.

**Comment by srikiz** [comment]
@johngmyers - Hopefully, I got it right :)
In my first commit, I just moved over the file from `pkg/resources/digitalocean/cloud.go` to `upup/pkg/cloudup/do/cloud.go` - with the same content.
In my second commit, I have modified the file to make it really work. 

All other changes are updating the dependencies to point to the new interface.
Hopefully this helps, please feel free to let me know if you need any additional changes to make it easier for review.

Thanks !

File: upup/pkg/fi/cloudup/do/cloud.go [hunk 1]
```diff
@@ -39,44 +38,64 @@ import (
...
+var _ fi.Cloud = &doCloudImplementation{}
+
 // Cloud exposes all the interfaces required to operate on DigitalOcean resources
-type Cloud struct {
+type doCloudImplementation struct {
 	Client *godo.Client
 
 	dns dnsprovider.Interface
 
 	// RegionName holds the region, renamed to avoid conflict with Region()
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go:84]: Should fix comment to match

File: upup/pkg/fi/cloudup/do/cloud.go.new [hunk 2]
```diff
@@ -0,0 +1,566 @@
+/*
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go.new]: This file should not be added then removed. Please edit it out of the commit stream.

File: protokube/pkg/protokube/do_volume.go [hunk 3]
```diff
@@ -128,9 +134,33 @@ func NewDOVolumes() (*DOVolumes, error) {
...
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+func NewDOCloud() (*godo.Client, error) {
+	accessToken := os.Getenv("DIGITALOCEAN_ACCESS_TOKEN")
+	if accessToken == "" {
+		return nil, errors.New("DIGITALOCEAN_ACCESS_TOKEN is required")
+	}
```
Comment (johngmyers) [inline protokube/pkg/protokube/do_volume.go:149]: This looks like a functional modification. I can't see any corresponding deletion.

**Review by johngmyers — CHANGES_REQUESTED** [review]
Unfortunately, the second commit is to big. I cannot easily follow to verify the lack of functional modifications. The fact that functions are being reordered at the same time as they or other things are being modified makes the diffs extremely hard to follow.

It also looks like the first commit won't compile, but is fixed up in the second commit. I can work around some of that, but it makes reviewing a harder task.

File: protokube/pkg/protokube/do_volume.go (hunk 3 above)
Comment (srikiz) [reply protokube/pkg/protokube/do_volume.go:149]: Protokube had a dependency on resources - digital ocean package and was getting a newCloud object by calling the dependency [here](https://github.com/kubernetes/kops/pull/11592/files?file-filters%5B%5D=.go#diff-c8f5df1bde4e360ef9070eb81ce136cc4d3c99220c0dc157300898b002826967L106)
Now, since that dependency is removed, it directly depends on the godo client and creates a new object here.

Comment (srikiz) [reply protokube/pkg/protokube/do_volume.go:149]: > Unfortunately, the second commit is to big. I cannot easily follow to verify the lack of functional modifications. The fact that functions are being reordered at the same time as they or other things are being modified makes the diffs extremely hard to follow.
> 
> It also looks like the first commit won't compile, but is fixed up in the second commit. I can work around some of that, but it makes reviewing a harder task.

@johngmyers - Apologize, its creating a problem in reviewing this. My first commit is failing since I just moved the same file over to upup/pkg/fi/cloudup/do. But I fixed all issues in my second commit, but placing of the functions didn't match well with the original code.

**Comment by srikiz** [comment]
> Unfortunately, the second commit is to big. I cannot easily follow to verify the lack of functional modifications. The fact that functions are being reordered at the same time as they or other things are being modified makes the diffs extremely hard to follow.
> 
> It also looks like the first commit won't compile, but is fixed up in the second commit. I can work around some of that, but it makes reviewing a harder task.

@johngmyers - Apologize, its creating a problem in reviewing this. My first commit is failing since I just moved the same file over to upup/pkg/fi/cloudup/do. But I fixed all issues in my second commit, but placing of the functions didn't match well with the original code.

**Comment by johngmyers** [comment]
(I haven't yet had time to review your latest push.)

@srikiz my primary task as reviewer is to verify the claim that there are no functional modifications. Like any human, I have limited short-term memory, so the fewer things I have to track at once the more practical the task.

So the request is to have a series of self-contained commits, so that I can review one commit, then discard most state about that commit and move to the next. As an exception to the expectation that each commit should be self-contained and compile, it is customary for larger changes to automatically generated files to be in a separate commit.

You might find it helpful to look at some of my refactoring PRs, for example #11219 (though that has functional changes mixed in).

File: upup/pkg/fi/cloudup/do/cloud.go (hunk 1 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go]: I have updated this now.

File: upup/pkg/fi/cloudup/do/cloud.go.new (hunk 2 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go.new]: Sure, Removed it.

**Comment by srikiz** [comment]
@johngmyers - thank you for your time on this. My updated commit stream should hopefully make it easier now. I have re-ordered the functions so it is in-line with the changes in the first commit. Please feel free to let me know if you need further modifications. Thanks again, for the help !!

File: protokube/pkg/protokube/do_volume.go (hunk 3 above)
Comment (johngmyers) [reply protokube/pkg/protokube/do_volume.go:149]: I can see the corresponding deletion now. Thanks.

File: upup/pkg/fi/cloudup/do/cloud.go [hunk 4]
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
...
-	return digitalocean.NewCloud(region)
+// TokenSource implements oauth2.TokenSource
+type TokenSource struct {
+	AccessToken string
+}
+
+// DOCloud exposes all the interfaces required to operate on DigitalOcean resources
+type DOCloud interface {
+	fi.Cloud
+	//Client() *godo.Client
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go:62]: nit: this should be removed

File: upup/pkg/fi/cloudup/do/cloud.go [hunk 5]
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
...
+	DropletsService() godo.DropletsService
+	DropletActionService() godo.DropletActionsService
+	VolumeService() godo.StorageService
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go:65]: Since this is specific to DO, perhaps it should stick with the DO terminology.

File: upup/pkg/fi/cloudup/do/cloud.go [hunk 6]
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
...
+// static compile time check to validate DOCloud's fi.Cloud Interface.
+var _ fi.Cloud = &doCloudImplementation{}
+
+// doCloudImplementation holds the godo client object to interact with DO resources.
+type doCloudImplementation struct {
+	Client *godo.Client
+
+	dns dnsprovider.Interface
+
+	// region holds the DO region
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go:85]: nit: should be a complete sentence, including punctuation:

```suggestion
	// region holds the DO region.
```

File: upup/pkg/fi/cloudup/do/cloud.go [hunk 7]
```diff
@@ -148,64 +168,69 @@ func (c *Cloud) DeleteInstance(i *cloudinstances.CloudInstance) error {
...
-	return c.RegionName
+func (c *doCloudImplementation) Region() string {
+	return c.region
 }
 
-// DNS returns a DO implementation for dnsprovider.Interface
-func (c *Cloud) DNS() (dnsprovider.Interface, error) {
-	return c.dns, nil
+func (c *doCloudImplementation) DNS() (dnsprovider.Interface, error) {
+	provider := dns.NewProvider(c.Client)
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go:187]: Why doesn't this return `c.dns`?

File: upup/pkg/fi/cloudup/do/cloud.go [hunk 8]
```diff
@@ -28,13 +42,526 @@ const TagNameEtcdClusterPrefix = "etcdCluster-"
...
+	return c.Client.Domains
+}
+
+func (c *doCloudImplementation) ActionsService() godo.ActionsService {
+	return c.Client.Actions
+}
+
+// FindVPCInfo is not implemented, it's only here to satisfy the fi.Cloud interface
+func (c *doCloudImplementation) FindVPCInfo(id string) (*fi.VPCInfo, error) {
+	return nil, errors.New("Not implemented")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/cloud.go:224]: Go errors should always start with lowercase letters

```suggestion
	return nil, errors.New("not implemented")
```

File: upup/pkg/fi/cloudup/do/mock_do_cloud.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+// Cloud exposes all the interfaces required to operate on DigitalOcean resources
+// type MockDOCloudService struct {
+// 	MockDroplets      godo.DropletsService
+// 	MockDropletAction godo.DropletActionsService
+// 	MockVolume        godo.StorageService
+// 	MockVolumeAction  godo.StorageActionsService
+// 	MockLoadBalancers godo.LoadBalancersService
+// 	MockDomain        godo.DomainsService
+// 	MockActions       godo.ActionsService
+// }
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/mock_do_cloud.go:43]: Remove commented-out code.

File: upup/pkg/fi/cloudup/do/mock_do_cloud.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+func (c *doCloudMockImplementation) DNS() (dnsprovider.Interface, error) {
+	provider := dns.NewProvider(c.Client)
+	return provider, nil
+}
+
+// FindVPCInfo is not implemented, it's only here to satisfy the fi.Cloud interface
+func (c *doCloudMockImplementation) FindVPCInfo(id string) (*fi.VPCInfo, error) {
+	return nil, errors.New("Not implemented")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/mock_do_cloud.go:71]: lowercase

File: upup/pkg/fi/cloudup/do/mock_do_cloud.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// DeleteGroup is not implemented yet, is a func that needs to delete a DO instance group.
+func (c *doCloudMockImplementation) DeleteGroup(g *cloudinstances.CloudInstanceGroup) error {
+	return fmt.Errorf("digital ocean cloud provider does not support deleting cloud groups at this time")
+}
+
+func (c *doCloudMockImplementation) DeleteInstance(instance *cloudinstances.CloudInstance) error {
+	return errors.New("Not tested")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/do/mock_do_cloud.go:80]: lowercase

**Review by johngmyers — APPROVED** [review]
These comments are all small enough to leave to follow-up work.

Thank you, that was much easier to review.

File: upup/pkg/fi/cloudup/do/cloud.go (hunk 4 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go:62]: Sure, updated.

File: upup/pkg/fi/cloudup/do/cloud.go (hunk 5 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go:65]: I'll probably keep this as-is for now, and rename this to StorageService as a separate PR.

File: upup/pkg/fi/cloudup/do/cloud.go (hunk 6 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go:85]: Sure, updated.

File: upup/pkg/fi/cloudup/do/cloud.go (hunk 7 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go:187]: Missed that, updated it.

File: upup/pkg/fi/cloudup/do/cloud.go (hunk 8 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/cloud.go:224]: sure, updated.

File: upup/pkg/fi/cloudup/do/mock_do_cloud.go (hunk 9 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/mock_do_cloud.go:43]: Sure, removed.

File: upup/pkg/fi/cloudup/do/mock_do_cloud.go (hunk 10 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/mock_do_cloud.go:71]: sure, updated.

File: upup/pkg/fi/cloudup/do/mock_do_cloud.go (hunk 11 above)
Comment (srikiz) [reply upup/pkg/fi/cloudup/do/mock_do_cloud.go:80]: Done.
